src/15-budget-intent/index.ts

```diff
@@ -0,0 +1,167 @@
+/**
+ * This experiment extracts a structured budget intent from user-supplied free text.
+ *
+ * Running this will generate multiple results across a number of examples and models.
+ *
+ * The results can be view in budgetIntents.ndjson
+ *
+ * They can also be converted to csv (via csv-kit) with:
+ *
+ * $ cat budgetIntents.ndjson | in2csv -f ndjson > budgetIntents.csv
+ *
+ * These^ csv results and some stats can be viewed at:
+ * https://docs.google.com/spreadsheets/d/1S1-7IJOPc26cgSlWaxP1bz9HEuNJXgQJTDdAyx7XRB0/edit?usp=sharing
+ *
+ */
+import _ from "lodash"
+import { anthropic } from "@ai-sdk/anthropic"
+import { LanguageModel, generateObject } from "ai"
+import { openai } from "@ai-sdk/openai"
+import { z } from "zod"
+import chalk from "chalk"
+import dedent from "dedent"
+import dotenv from "dotenv"
+import fs from "fs"
+import pAll from "p-all"
+import pSeries from "p-series"
+
+dotenv.config()
+
+const MODELS: Record<string, LanguageModel> = {
+  turbo: openai("gpt-3.5-turbo"),
+  fourO: openai("gpt-4o"),
+  haiku: anthropic("claude-3-haiku-20240307"),
+  // sonnet: anthropic("claude-3-sonnet-20240229"),
+  sonnet35: anthropic("claude-3-5-sonnet-20240620"),
+  opus: anthropic("claude-3-opus-20240229"),
+}
+
+const EXAMPLES = [
+  "I’d spend up to $100,000 this year on my collection.",
+  "I have $30,000 left in my decorating budget and I’d like to fill the walls with art.",
+  "I will spend $500 - $1,000 on this artwork.",
+  "none of your business",
+  "under 1k",
+  "1000 quid",
+  "I have 10k to spend on two works",
+]
+
+const chalkFn: Record<string, (...args: string[]) => void> = {
+  turbo: chalk.green,
+  fourO: chalk.green.dim,
+  haiku: chalk.blue,
+  opus: chalk.blue.dim,
+  sonnet: chalk.cyan,
+  sonnet35: chalk.cyan.dim,
+}
+
+const system = dedent`
+  Your task is to analyze a plaintext budget statement
+  and produce a structured object to describe it.
+
+  If a budget expresses a range, set min and max accordingly.
+
+  If a budget expresses a single amount, then set min to 0 and max to the amount.
+
+  If a budget cannot be inferred, set both min and max to -1.
```

Comment:
I took this expedient route for a budget than can't be inferred. Consumers will basically check `if (intent.budget.max > 0) { doSomething() }` or something like that.